### ISSUE TITLE
MP: Friends cannot see AD Courses/Waypoints in local Server

### ISSUE THREAD

**Issue by Blithe89** [Stephan-S/FS19_AutoDrive#1088]
AutoDrive Version: 1.0.7.0-37
Release or custom zip: both
Map (if applicable): All (i guess)
Vehicle / type (applicable): All

Deutsche Übersetzung kann ich machen, wenn Bedarf besteht. 

**Describe the bug**
When I play with my friends on my local save (my computer is host and client), my friends cannot see
the AD courses. I am the only one who can use the courses for the map (Multimap2019) from another user. 

I checked out a lot of videos, read a lot and tried out, but I couldnt find a way to sync my waypoint to the others. So maybe its not really a bug, I dont know.

A bit more detail:

-I found a pack for the map, where were all courses allready included, so no driving the courses manually.
-I put the config.xml with the courses in my savegamefolder from the game. 
-I started the server (always local(on my pc) MP, never SP)
-It worked, everything fine.
-A friend came, he couldnt see any courses, than I went into a vehicle (maybe he also in the moment (not clear if it was the same time, or one was earlier than the other)) and a calculating began (yellow line in the middle lower part of screen)
-after this, he could also see all the courses. Perfect.
-after I started the server again and friend joined, he couldnt see courses again. Went to the vehicles, nothing happend. We pushed the update button in the editormenu (yellow calculating in the lower part again). First me, than he, than me, nothing happend. I am still the only one who can see the courses.

I (and my friends) updated the AD from 36 to 37 nothing changed.
But I saw some pre AD courses in the mod. So maybe there is a possiblity to get the courses from the mod.zip and not over the Savegame. So I decided to try it out.

This was MP alone:
I loaded one map (Eiersholt) which is allready in the Autodrive mod and all courses was there from the beginning without I am touching the savegame.
I tried it out to put my config from the multimap to the others configs and start a new map. No courses. I changed the config to AutoDrive_Multimap_2019_init_config.xml with init inside, started a new map, It worked. Everything (courses) there. 

-I give this "new" FS19_Autodrive.zip to my friends and we started our normal map (I deleted the config in my Savegame folder to be sure, that it is loading the config new from the zip)
-I saw the courses again, the others not. We pushed the update button, nothing helped.

The intresting part for me is: if they open the editor in the AD hud (the small one on the right side) they can see all the lines on the streets, but no waypoints (also not on the map). So also no dropdown menu to choose a target for the vehicle. Somehow they received the information partly.

I am sorry if I oversaw a kind of instruction for this problem. But what can I do, that I am not the only person who can see the courses on our map? The savegame there all the courses are, is not there on the other computers. So how to transmitter the courses?

I hope it was understandable.

Anyways, even if I am the only one who can use this mod on our server, this mod is really helpfull. Thank you for all the time you spent allready into this mod and probably also in the future.

**Comment by Iwan1803** [comment]
Same here, but only on big maps with a big AD route config file.
It takes a very long time till the routes are synched and then may be or not AD shows them.

**Comment by Blithe89** [comment]
Well, this config file I use is around 60mb. And the Multimap2019 is 4x. I guess it belongs to the bigger ones.
Is it kind of coincidence if this (not) happend, or are there some recommends, for example: "shouldnt bigger than 10mb, or not more than 50 waypoints"?

How would be a normal sync? Everybody goes on server and it automaticly makes a sync, or the host has to push the recalculating button, or the clients? This is the point I couldnt find anywhere.

**Comment by Iwan1803** [comment]
We tried it with a ~6 MB file. After joining the server and after the normal synch process i jumped in a tractor. Then at the top of the AD panel i could see the the start of the sync progress.
The first on the server (after the host) got a full synch and everything works. For the 2nd it's like gambling.
Recalculation doesn't solve the problem and i can't see a shortcut for a re-synch.
I don't know if this is only a problem, if someone is hosting a game or if it is running on a dedicated server.

**Comment by Blithe89** [comment]
Puh, if allready 6mb is "too much"...

So the sync process is normally automatic. Would be nice, if somebody share his/her opinion/experience from a dedicated server point of view. Working more stable there?

**Comment by Axel32019** [comment]
From my point of view the issue is that still the classic calculated route data is active. 
This means the calculated routes, which are not used in live-mode, are still stored in XML, synchronized in MP etc.
The amount of data grows with the size of the map (waypoints) and thefore also with the number of targets:
data = number of waypoints * number of targets
If someone is interested I can remove the sync, but I have no MP to test it.
@Blithe89 Are you able to exchange single file(s) in the Mod? If yes I can add it here...

**Comment by Stephan-S** [comment]
The problem should not be with the navigational data. This is only a portion of the transmitted data.
We still have to transmit the id and the positions of the waypoints. So this won't change much.

Most edit actions you can do are now happening without transferring much data anyway.
And if a whole reload of the network is required, it will be displayed during that phase in the HUD, so you probably won't be surprised if something doesn't show up during that time. So I'm leaning to the fact that there is some bug in the codebase which prevents proper syncing in certain use cases.

Can you please provide the log files? Of course one, after a second or third player has joined and doesn't receive any data.
For the syncing to start, you have to be inside a vehicle.
And again, this syncing process should be displayed with text and a growing percentage inside the AD Hud.

**Comment by Blithe89** [comment]
Thank you all for reply.

@Axel32019 I didnt get the point with the live-mode. The routes which didnt used in this time after reboot server in the live-mode, cannot be snychronized? Or if the sync at the beginning is off, than it can actually transmitted? Sorry, I am confused.

And yes, I can exchange single files. (I did it with the configs)

@Stephan-S  Today I can only one other player. Should I do something special in the procedure? Turning some things off in the deeper menu (or on?)?
Who has to start the sync? I (server), or client in the vehicle? (or doesnt matter, because the calculating is on my computer anyways?)

The part with the growing percentage is the yellow one in the lower middle part of the screen I guess. 



Wird wahrscheinlich aber erst nach 23Uhr mit dem Log.

**Comment by Axel32019** [comment]
I would say the navigation data is the much bigger part of all data.
For each waypoint you have:
- ID, x, y, z, out, in -> 6 values minimum, plus 1... for branch(es) if there are any
- 200, 300,... values for the navigation (targets), as I could not see that the "compression" like in the XML file is not applied, or am I wrong?

This is only just a rough calculation.

@Blithe89 If you want you can try the attached file.
[AutoDriveEventUtil.zip](https://github.com/Stephan-S/FS19_AutoDrive/files/4200771/AutoDriveEventUtil.zip)
In live-Mode the calculated navigation data is not used to find a way to the target, so there is also no need to synchronise this from my understanding.

**Comment by Blithe89** [comment]
@Axel32019  should I put this file only on the server-client (host), or also on the other clients too? (Dont know if the clients take the information from the servers). 
But even if you dont calculate again you need the waypoints which you have to sync first. In SP there is no problem. Its done in the savegamefolder. But in MP only the serverclient has this. So the waypoints(destinationpoints) and courses(the single points on the street) must be transmitted first before you can block the sync. And in our case it seems the course is transmitting (clients can see the lines on the street, but nothing to choose from the destinationmenu (destinationpoints are missing).

But I will try out the file. Feedback for this I cannot give before next week (normally no sessions on weekend^^)


@Stephan-S The questions is: should the sync process triggered alone if you are enter a vehicle, or then you have the possibility to push the updatebutton in the AD HUD and then the sync start. Maybe I oversaw it, but normally I only see the calculation in the middle of the screen (lower part) not in the HUD (but I didnt have a look, I will check this later again)


Here the logs from me. Log2 is only the information which could be intresting I guess.
Log is normal (except IDs).
[log.txt](https://github.com/Stephan-S/FS19_AutoDrive/files/4204494/log.txt)
[log2.txt](https://github.com/Stephan-S/FS19_AutoDrive/files/4204495/log2.txt)



(by the way: the most of the warning are from GC icons. This is new since yesterday, because I put the mod inside yesterday. Has nothing to do with problem (it exist before already))


@all is there any gerneral settings which is important or you should touch, for example the livemode on/off?

Thank you anyway for taking the time allready.

**Comment by Axel32019** [comment]
@Blithe89 You should exchange the file in all used (!!!) AD Mods, on server and all clients. 
What you mean with " Its done in the savegamefolder"? The file is for the Mod, not the savegame (AutoDrive_..config.xml)!
The waypoints on the street are sync first, the destinations last. That might be the reason why you see the course on the street.
To get it more clear could you add the AutoDrive_...config.xml file in a .zip here? Or write some numbers out of it?
I am interested in:
- last number in line `"<id>"`
- last numer of `"<mm...>"`

**Comment by Iwan1803** [comment]
@Axel32019 
wenn man einem Server joined, wird die Datei mit den Wegpunkten doch automatisch synchronisiert. Keine Ahnung wohin er sie dann schreibt, aber nicht in den Mod selber noch in einen Savegame-Ordner oder irre ich mich? In den AD-Import & -Export-Ordner habe ich auch nichts gefunden.

**Comment by Axel32019** [comment]
@Iwan1803 Im zip ist eine LUA die die Navigationsdaten nicht synchronisiert, es geht nicht um die AutoDrive_..config.xml. Mich interessieren aus der AutoDrive_..config.xml nur die Anzahl der Wegpunkte und die Anzahl der Ziele. Daraus ergibt sich wie oben erwähnt die Datenmenge der im Live-Modus nicht mehr benötigten zu synchronisierenden Daten - da sehe ich das Optimierungspotential. Vieleicht wurde das bisher missverstanden :-(

**Comment by Iwan1803** [comment]
Wenn die ganzen Wegpunktdaten aber im MP nicht mehr mit dem Server (Dedi oder einfacher Host) synchronisiert werden, wo sollen die Spieler sie dann herkommen, um damit weiter arbeiten zu können?

**Comment by Axel32019** [comment]
Es geht nicht um die Wegpunkte, sondern um die Daten die ohne Live-Modus berechnet werden mussten. Siehe vorher: ID, x, y, z, out, in und die Ziele werden ja immer noch synchronisiert, mehr braucht es nicht.

**Comment by Stephan-S** [comment]
Eigentlich hätten die Navigationsdaten ja noch nie übertragen werden müssen, wenn ich so recht darüber nachdenke. Da die Berechnung der Strecke ja eh immer schon auf dem Host ausgeführt wurde und dann nur die geplante Strecke auch bei den Clients landen sollte.
Also können wir deine EventUtil Optimierung so oder so übernehmen denke ich. Ich schau mir das noch mal an.

@Blithe89 The (re)calculation of the paths is completely independent from the synchronization with the players. The sync should take place when someone enters a vehicle and a certain timer (few seconds) is run out (to prevent too much traffic right at the start of the game session).
The percentage message I am talking about is displayed at the top bar of the HUD where the version number is displayed. This is not the yellow recalculation message.

But after a recalculation, you trigger a new complete sync process. 
Each course edit action, such as creating or removing connections/nodes etc. should be transmitted as event immidiately.

So each joining player should see the sync messages and only after this is completed, they can see all the destinations. Since they are transmitted last, as @Axel32019 already said.

So please wait for a few second and check if the sync has started/completed for each player or if anything has prevented this from happening.

Please report back if you have new information. Thank you

**Comment by Axel32019** [comment]
@Stephan-S Ich hab da noch ein paar mehr Ideen zur Optimierung. Sollten wir die in einem eigenen Issue mal sammeln / diskutieren?

**Comment by Stephan-S** [comment]
@Axel32019 klar gerne. wir könnens uns aber auch gerne kurz im Discord zusammen setzen. Ich bin gerade dran deinen Live Mode nun endlich zur einzigen Möglichkeit zu machen und die alte Berechnung rauszunehmen.

**Comment by Blithe89** [comment]
Hi,

ich dachte das der Input/Output wohl höher wäre, wenn man versucht englisch zu schreiben. Aber da jetzt klar, dass wirklich jeder der aktuell Beteiligten hier deutsch ist, können wir ja erstmal in der Muttersprache weitermachen. Eventuell eine Lösung, wenn nötig, dann nochmal englisch erstellen.

@Axel32019 [die config die ich habe, ist fast die gleiche wie von dem Threadersteller Spielera paar Tage später. Aufgrund der "Rechtschreibfehler" und des Aufbaus, ist sie auch vom gleichen Autor. Auch bei mir ist sie fast 60mb groß. Die Felder wurden immer komplett umfahren, was wohl helfen soll das Abfahren von den Feldern zu erleichtern, da der kürzeste Punkt zum Streckensystem genutzt wird, aber keine Ahnung ob das so sinnvoll ist. Aber komplett alles neu einzufahren ist dann auch übel. Was scheinbar nur einer gleichzeig machen kann. Am besten der Host]
Zur config selbst:
[AutoDrive_Multimap_2019_config.zip](https://github.com/Stephan-S/FS19_AutoDrive/files/4209684/AutoDrive_Multimap_2019_config.zip)
Die letze ID wäre 35893
und die mm wäre 291

(Mit meinen Verständniss kann man jetzt aufgrund dieser beiden Werte nicht gleich was berechnen, da das ja Maximalwerte sein drüften. Sprich wenn Punkte im Spiel wieder rausgelöscht werden, müsste die ID und die mm trotzdem fortlaufen, nur die entsprechenden gleöschten Punkte fehlen in der Config (oder werden negiert?))

Zu deiner Lua etwas weiter unten.

@Stephan-S da bisher keine Session zustande gekommen ist und ich nur im Multiplayer überprüfen kann, werde ich später berichten, ob es besser geworden ist. 
Ich habe die Lua von Axel32019 eingefügt in meiner Zip, aber du hast ja auch die Mod geupdated. Ich geh davon aus, dass die Veränderung der Lua da mit drin ist und es sinnvoller wäre, gleich deine neue Version zu nehmen. 

So wie ich das sehe:
wenn ein Client den Server betritt, werden bei einen Fahrzeugbeitritt die Wegpunktedaten synchronisiert. Aber keine Routen. Das macht der Livemodus, wenn der entsprechenden Client das entsprechende Ziel auswählt und auf AD starten im HUD drückt.

Da einige vom starken Ruckeln während der AD Fahrt berichten, könnte ich mir vorstellen, dass der Livemode bei jeden neu ereichten Punkt auf der Straße eine Neuberechnung zum Ziel vornimmt, oder? Wenn dem so ist, wäre es doch vielleicht besser den Livemode so zu verändern, das er die Route nur einmal bei Startbutton berechnet und dann die Route temporär (solange bis Ziel erreicht ist, oder AD abgestellt wurde) beim Clienten speichert.

**Comment by Stephan-S** [comment]
Ja, du kannst die aktuelle Entwicklerversion nehmen, dort sind die Änderungen bereits eingetragen und auch deine 60MB Datei wird damit zu 1,6MB zusammen gestampft.

Der Live-Modus berechnet natürlich nur einmal die Route, nicht bei jedem einzelnen Wegpunkt.
Allerdings wird bei Erreichen eines Wegpunkts jeweils ein UpdateEvent vom Server an alle Spieler gesendet, das den entsprechenden Zähler hochzieht, um beispielsweise auch die Anzeige der nächsten beiden Streckenpunkte (weiße Linie) zu synchronisieren. Das ist aber in Normalfall kein großes Thema und sollte an sich nicht zum Ruckeln führen.

Wenn du dir genauer ansehen willst, was alles gesender wird, kannst du in den Debugeinstellungen mal das Häkchen bei 'Netzwerk' setzen. Dann wird jedes Event ins Log geschrieben. Wenn ihr Ruckler habt, kannst du hier ja mal ein entsprechend aufgenommenes Log posten und wir schauen mal woran es liegen könnte

**Comment by Blithe89** [comment]
@all Wir konnten die neuere (40glaube, näheres dazu unten) Version testen und es verlief äußerst positiv. 

Wir waren insgesamt drei Leute und bei jeden funktionierte es problemlos und beim ersten Versuch. 
Die Synchroinfo habe ich als Host zwar nicht gesehen (extra drauf geachtet), aber beim Clienten war sie da. Ging auch sehr schnell (paar Sekunden), ohne Ruckeln, etc.
(nicht wie courseplay, aber das ist eine andere Geschichte...)

Ich finde es auch gut, das diverse Einstellmöglichkeiten verschwunden sind. So kommt man garnicht in die Versuchung Problem zu erzeugen (z.b. Livemode)


Ich hätte da noch eine Anmerkung. Die Modversion (auch Dev) besser kennzeichnen. Wenn ich meinen Config von der Map ansehe, steht da 40. Wenn ich die modDesc(ription?) ansehe, steht da eine 46.

Jetzt hatten wir gestern einen Issue, wo nicki0705 ein Problem beschreibt, welches mit diesem hier nahezu identisch ist. Es wurde geschrieben Mod ist aktuelle Version. Welche ist das aber? Die Ereignisse haben sich ja grade etwas überschlagen und das neue Verfahren ist ein Segen für MP.
Also richtig super!
Vielleicht hat nicki einfach nur einen veralteten (paar Tage) Mod...

**Comment by TyKonKet** [comment]
**Hello, Thank you for opening this issue!**
Since we have just released [the new refactored version of AutoDrive](https://github.com/Stephan-S/FS19_AutoDrive/releases/tag/1.1.0.0), we can no longer provide any feedback or support for open bug reports related to the old code base.
If the problem persists, please reopen a new issue with all the required details.
Thank you for understanding and we hope you continue enjoying our mod 😄
